packages/core/src/index.ts

```diff
@@ -0,0 +1,144 @@
+import { BlockAddedEvent, EditorJSModel, EventType } from '@editorjs/model';
+import type { CoreConfig, CoreConfigValidated } from './entities/Config.js';
+import { composeDataFromVersion2 } from './utils/composeDataFromVersion2.js';
+import ToolsManager from './tools/ToolsManager.js';
+import { BlockToolAdapter, CaretAdapter } from '@editorjs/dom-adapters';
+import type { BlockAPI, BlockToolData, API as EditorjsApi, ToolConfig } from '@editorjs/editorjs';
+import type { BlockTool } from './entities/BlockTool.js';
+
+/**
+ * Editor entry poit
+ * - initializes Model
+ * - subscribes to model updates
+ * - creates Adapters for Tools
+ * - creates Tools
+ * - adds Blocks accodring to model updates
+ */
+export default class Core {
+  /**
+   * Editor's Document Model
+   */
+  #model: EditorJSModel;
+
+  /**
+   * Tools manager is responsible for creating tools
+   */
+  #toolsManager: ToolsManager;
+
+  /**
+   * Editor configuration
+   */
+  #config: CoreConfigValidated;
+
+  /**
+   * @param config - Editor configuration
+   */
+  constructor(config: CoreConfig) {
+    this.validateConfig(config);
+    this.#config = config as CoreConfigValidated;
+
+    this.#toolsManager = new ToolsManager(config.tools);
+
+    const { blocks } = composeDataFromVersion2(config.data ?? { blocks: [] });
+
+    this.#model = new EditorJSModel();
+    this.#model.addEventListener(EventType.Changed, (event: ModelEvents) => this.handleModelUpdate(event));
+    this.#model.initializeDocument({ blocks });
+  }
+
+  /**
+   * Validate configuration
+   * @param config - Editor configuration
+   */
+  private validateConfig(config: CoreConfig): void {
+    if (config.holder === undefined) {
+      const holder = document.getElementById('editorjs');
+
+      if (holder) {
+        config.holder = holder;
+      } else {
+        throw new Error('Editor configuration should contain holder or #editorjs element should be present');
+      }
+    }
+
+    if (config.data) {
+      if (config.data.blocks === undefined) {
+        throw new Error('Editor configuration should contain blocks');
+      }
+
+      if (!Array.isArray(config.data.blocks)) {
+        throw new Error('Editor configuration blocks should be an array');
+      }
+    }
+  }
+
+  /**
+   * When model emits block-added event, add an actual block to the editor
+   * @param event - Any model event
+   */
+  private handleModelUpdate(event: ModelEvents): void {
+    if (event instanceof BlockAddedEvent === false) {
+      return;
+    }
+
+    void this.handleBlockAdded(event);
+  }
+
+  /**
+   * Insert block added to the model to the DOM
+   * @param event - Block added event
+   */
+  private async handleBlockAdded(event: BlockAddedEvent): Promise<void> {
+    /**
+     * @todo add batch rendering to improve performance on large documents
+     */
+    const index = event.detail.index;
+
+    if (index.blockIndex === undefined) {
+      throw new Error('Block index should be defined');
```

Comment:
If blockIndex is not there that means something is wrong with the model. This error would give a user no information on how to fix that